kratos/linear_solvers/fallback_linear_solver.h

```diff
@@ -0,0 +1,897 @@
+        while (!success && mCurrentSolverIndex < mSolvers.size()) {
+            success = GetCurrentSolver()->Solve(rA, rX, rB);
+            // In case of failure
+            if (!success) {
+                // First, update the counter
+                UpdateCounterSolverIndex();
+
+                // Call initialize methods
+                InitializeSolutionStep(rA, rX, rB);
+            }
+        }
+        return success;
+    }
+
+    /**
+     * @brief Multi solve method for solving a set of linear systems with same coefficient matrix.
+     * @details Solves the linear system Ax=b and puts the result on SystemVector& rX. rX is also th initial guess for iterative methods.
+     * @param rA. System matrix
+     * @param rX. Solution vector. it's also the initial
+    guess for iterative linear solvers.
+     * @param rB. Right hand side vector.
+     */
+    bool Solve(
+        SparseMatrixType& rA,
+        DenseMatrixType& rX,
+        DenseMatrixType& rB
+        ) override
+    {
+        if (mResetSolverIndexEachTry) mCurrentSolverIndex = 0;
+        bool success = false;
+        while (!success && mCurrentSolverIndex < mSolvers.size()) {
+            success = GetCurrentSolver()->Solve(rA, rX, rB);
+            // In case of failure
+            if (!success) {
+                // First, update the counter
+                UpdateCounterSolverIndex();
+
+                // // Call initialize methods (NOTE: does not exist the method for dense matrices)
+                // InitializeSolutionStep(rA, rX, rB);
+            }
+        }
+        return success;
+    }
+
+    /**
+     * @brief Eigenvalue and eigenvector solve method for derived eigensolvers
+     * @param K The stiffness matrix
+     * @param M The mass matrix
+     * @param Eigenvalues The vector containing the eigen values
+     * @param Eigenvectors The matrix containing the eigen vectors
+     */
+    void Solve(
+        SparseMatrixType& K,
+        SparseMatrixType& M,
+        DenseVectorType& Eigenvalues,
+        DenseMatrixType& Eigenvectors
+        ) override
+    {
+        KRATOS_ERROR << "This is designed for linear system only, not for eigen values calculation" << std::endl;
+    }
+
+    /**
+     * @brief Checks if additional physical data is needed by the solver.
+     * @details Some solvers may require a minimum degree of knowledge of the structure of the matrix. 
+     * For instance, when solving a mixed u-p problem, it is important to identify the row associated with v and p.
+     * Another example is the automatic prescription of rotation null-space for smoothed-aggregation solvers,
+     * which require knowledge of the spatial position of the nodes associated with a given degree of freedom (DOF).
+     * @return True if additional physical data is needed, false otherwise.
+     */
+    bool AdditionalPhysicalDataIsNeeded() override
+    {
+        return GetCurrentSolver()->AdditionalPhysicalDataIsNeeded();
+    }
```

Comment:
this is incorrect

Imagine the preferred solver doesn't require extra stuff (e.g.: CG) but the second in line does (e.g.: AMG). When B&S checks `FallbackLinearSolver::AdditionalPhysicalDataIsNeeded` it will return `false` (because CG is the current solver) and never call `AMGCLSolver::ProvideAdditionalData`. However, it CG fails and moves on to AMG, it wont't be properly set up and will probably do something nasty.

I'm not sure how to solve this though. Calling `AMGCLSolver::ProvideAdditionalData` in this scenario before CG fails is definitely a waste.

Reply:
The problem is that you cannot call in every solver this method because it may be incosistent. Some times this method modifies the LHS and RHS. So, you cannot called in both. This is a complex discussion, we disacussed this inernally and this was the most consistent solution.

Reply:
> we disacussed this inernally and this was the most consistent solution

That doesn't help me because I don't have access to your internal discussions.

I understand if you don't want to support this, but then you need to detect when it happens and throw an exception that explains exactly what went wrong. Also, it'd be nice if this caveat was mentioned in the docs (preferably highlighted with `@note` or `@warning`).

Reply:
> This is a complex discussion, we disacussed this inernally and this was the most consistent solution.

that does not help ppl outside your group. Please give a brief summary

Reply:
You don't have acess to our discussion, and this was a different function, I got confused with so many comments. What can be done is to compute an OR